### ISSUE TITLE
Pluck last token before lm_head(x) during inference?

### ISSUE THREAD

**Issue by jxtps** [karpathy/nanoGPT#42]
Reading the [forward code](https://github.com/karpathy/nanoGPT/blob/master/model.py#L136), the [generate code](https://github.com/karpathy/nanoGPT/blob/master/model.py#L263), and the [training code](https://github.com/karpathy/nanoGPT/blob/master/train.py#L93) it looks like you only need all logits during training, and that during generation (inference) only the last logit is actually used. 

Since `lm_head` is a substantial op, it could make sense to pluck the last token **before** `lm_head` as opposed to after during inference? 

I.e. something like: 
```python
# if we are given some desired targets also calculate the loss
loss = None
if targets is not None:
    logits = self.lm_head(x)
    loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1)
else:
    logits = self.lm_head(x[:, :-1, :]).squeeze(1)

return logits, loss
```
*Haven't tested that code, apologies for any typos/errors.*

**Comment by karpathy** [comment]
I don't believe this saves tooooo much time because the lm_head is pretty fast. Alos your indexing above is wrong. But I pushed this change for now anyway, ty!: https://github.com/karpathy/nanoGPT/commit/8f85b8334749d1b6c3af3e11f4874bc1d0c0cad2

**Comment by jxtps** [comment]
Awesome! Another micro-optimization: the bias [is partially recomputed in each attention module's forward pass](https://github.com/karpathy/nanoGPT/blob/master/model.py#L25): 
```python
def __init__(self, config):
    ...
    self.register_buffer("bias", torch.tril(torch.ones(config.block_size, config.block_size))
                                .view(1, 1, config.block_size, config.block_size))
def forward(self, x):
    ...
    att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))
```
If you change that to something like: 
```python
    self.register_buffer("bias", torch.tril(torch.ones(config.block_size, config.block_size))
                                .view(1, 1, config.block_size, config.block_size) == 0)
...
    att = att.masked_fill(self.bias[:,:,:T,:T], float('-inf'))
```
it should save one GPU op, which used to have some overhead but may not matter anymore with `torch.compile`?

(and yes, the bias could be shared across attention modules, but I agree, that would introduce more complexity than it's worth) 

Ok, I'll see myself out now, fantastic project, thanks for sharing it! ;)
